### PR TITLE
Improve enrich mediator jsonpath flow

### PR DESCRIPTION
**Pull request by arunans23**
Currently there are some limitations in enrich mediator jsonpath flow. This PR will resolve it and support enriching jsonObjects to Properties as well.

Related to https://github.com/wso2/product-ei/issues/3430

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: modules/core/src/main/java/org/apache/synapse/mediators/eip/EIPUtils.java [hunk 1]
```diff
@@ -321,4 +330,36 @@ public static Object formatJsonPathResponse(Object input) {
...
+     * @param primaryPayload The json object where the key value pairs will be added
+     * @param secondaryPayload The json object whose key value pair will be added to primaryPayload
+     */
+    public static void mergeJsonObjects(JsonObject primaryPayload, JsonObject secondaryPayload) {
+        for (String key : secondaryPayload.keySet()) {
+            primaryPayload.add(key, secondaryPayload.get(key));
+        }
+    }
+
+    // Set default configuration for Jayway JsonPath
```
Comment (milindaperera) [inline modules/core/src/main/java/org/apache/synapse/mediators/eip/EIPUtils.java]: Let's add Javadoc

File: modules/core/src/main/java/org/apache/synapse/mediators/elementary/EnrichMediator.java [hunk 2]
```diff
@@ -91,11 +96,48 @@ public boolean mediate(MessageContext synCtx) {
...
+
+        //A JsonElement to store the parsed json object, so that it can be used without parsing the string continuously
+        JsonElement sourcePropertyJson = null;
+
+        /*
+        * Here we are checking whether the source Property is of XML type,
+        * so that we can decide on the flow (Jsonpath or xpath)
+        * A value stored in a property can be of type ArrayList, OMElement or String
+        */
+        boolean isSourcePropertyXML = false;
```
Comment (milindaperera) [inline modules/core/src/main/java/org/apache/synapse/mediators/elementary/EnrichMediator.java]: Let's reformat

File: modules/core/src/main/java/org/apache/synapse/mediators/elementary/Target.java [hunk 3]
```diff
@@ -421,10 +474,15 @@ private void setEnrichResultToBody(MessageContext synapseContext, SynapseJsonPat
         String newJsonString = "";
 
         if (action.equalsIgnoreCase(ACTION_REPLACE)) {
-            // replaces an existing value in json
-            newJsonString = JsonPath.parse(jsonString).set(expression, sourceNode).jsonString();
+            if (isRootPath) {
+                log.error("Acting replace is not supported for root path in type custom. " +
```
Comment (milindaperera) [inline modules/core/src/main/java/org/apache/synapse/mediators/elementary/Target.java]: Shall we check the possibility to move this validation to deployment

File: modules/core/src/main/java/org/apache/synapse/mediators/elementary/EnrichMediator.java (hunk 2 above)
Comment (arunans23) [reply modules/core/src/main/java/org/apache/synapse/mediators/elementary/EnrichMediator.java]: Done.

File: modules/core/src/main/java/org/apache/synapse/mediators/eip/EIPUtils.java (hunk 1 above)
Comment (arunans23) [reply modules/core/src/main/java/org/apache/synapse/mediators/eip/EIPUtils.java]: Done

File: modules/core/src/main/java/org/apache/synapse/mediators/elementary/Target.java (hunk 3 above)
Comment (arunans23) [reply modules/core/src/main/java/org/apache/synapse/mediators/elementary/Target.java]: Done

File: modules/core/src/main/java/org/apache/synapse/config/xml/EnrichMediatorFactory.java [hunk 4]
```diff
@@ -227,6 +226,13 @@ private void populateTarget(Target target, OMElement sourceEle) {
                 } catch (JaxenException e) {
                     handleException("Invalid XPath expression: " + xpathAttr);
                 }
+                if (target.getAction().equals(Target.ACTION_REPLACE) && (target.getXpath() instanceof SynapseJsonPath)
+                        && ("$".equals(((SynapseJsonPath) target.getXpath()).expression) ||
```
Comment (sdkottegoda) [inline modules/core/src/main/java/org/apache/synapse/config/xml/EnrichMediatorFactory.java]: Shall we extract target.getXpath as targetXPath and target.getXpath()).expression as targetExpression for better clarity?

Comment (arunans23) [reply modules/core/src/main/java/org/apache/synapse/config/xml/EnrichMediatorFactory.java]: Fixed in https://github.com/wso2/wso2-synapse/pull/1445

**Pull request merged**